### DISCUSSION TITLE
Unsure how to build/link imgui as a library in cmake [SOLVED]

### DISCUSSION THREAD

**Discussion by IssaAboudi** [ocornut/imgui#4915] [New Users Build/Link/Run issues ONLY!]
Hello everyone,

I am really new to cmake and I'm not understanding how to build/link imgui to my project. I've provided my CMakeLists.txt file here:

```
cmake_minimum_required(VERSION 3.21)
project(learning)

set(CMAKE_CXX_STANDARD 14)

set(IMGUI_PATH  "~/lib/imgui/")

file(GLOB IMGUI_SOURCES ${IMGUI_PATH}/*.cpp)
add_library(ImGui STATIC ${IMGUI_SOURCES})
target_include_directories(ImGui PUBLIC ${IMGUI_PATH})

add_executable(learning main.cpp)

target_link_libraries(learning PUBLIC ImGui)

```

I have been able to import all the files into my project in Clion and include "imgui.h" in main, but I get undefined references. When I start adding the cpp to the `add_executable()`, it starts giving me more and more undefined references:

To recreate this error, I changed CMakeLists.txt to:
```
cmake_minimum_required(VERSION 3.21)
project(learning)

include_directories(imgui)

add_executable(learning main.cpp imgui/imgui.cpp)
```
![image](https://user-images.githubusercontent.com/65807151/149643911-8d8d47e6-04a8-4328-afac-9373f4081dea.png)

Any help would be greatly appreciated. I don't really understand cmake and guides online have been difficult to follow.

**Comment by PathogenDavid** [comment]
Your first `CMakeLists.txt` looks correct to me.

Your second `CMakeLists.txt` is failing because you're only including `imgui.cpp`, you need `imgui_demo.cpp`, `imgui_draw.cpp`, `imgui_tables.cpp`, and `imgui_widgets.cpp` too.

It's not really clear to me what the relationship between the two different `CMakeLists.txt` files is.

**Reply by IssaAboudi** [reply]
Hey sorry about the late response, I didn't see this.

I'm still getting a lot of undefined references with the first one and I'm not sure what else to adjust.

My current cmakelists.txt looks like:

```cmake_minimum_required(VERSION 3.21)
project(learning)

set(CMAKE_CXX_STANDARD 14)

set(IMGUI_PATH  "~/lib/imgui/")

file(GLOB IMGUI_SOURCES
        "~/lib/imgui/*.h"
        "/imgui/*.cpp")
add_library(app ${IMGUI_SOURCES} imgui/imgui.cpp)

add_executable(learning main.cpp)

target_link_libraries(learning PUBLIC app)
```
I'm having dozens of undefined references and I'm not sure how to specify where cmake should look for their definitions. I am including glfw3.h and its finding the headers okay

**Reply by IssaAboudi** [reply]
I've chatted with some people and figured out my problem - I started adding things one by one to cmake and once I did that, I fixed the undefined references error

My final CMakeLists.txt (for others who may need help):
```cmake_minimum_required(VERSION 3.21)
project(TestImGui)

set(CMAKE_CXX_STANDARD 14)

#fetched these using vcpkg manager
find_package(glfw3 CONFIG REQUIRED) #got by doing 'vcpkg install glfw'
find_package(glad CONFIG REQUIRED) #got by doing 'vcpkg install glad'

#make an imgui folder and put all of these files in the imgui folder
set(IMGUI_SRC
        imgui/main.cpp
        imgui/imgui_impl_opengl3.h
        imgui/imgui_impl_opengl3.cpp
        imgui/imgui_impl_opengl3_loader.h
        imgui/imgui_impl_glfw.h
        imgui/imgui_impl_glfw.cpp
        imgui/imgui.h
        imgui/imgui.cpp
        imgui/imgui_draw.cpp
        imgui/imgui_tables.cpp
        imgui/imgui_widgets.cpp
        imgui/imgui_demo.cpp
        )

#Create the executable and link against all the files provided to us by ImGui
add_executable(TestImGui ${IMGUI_SRC})

target_link_libraries(TestImGui PRIVATE glfw glad::glad) #link the two libraries into the application
```

This is what my project directory looks like:
![image](https://user-images.githubusercontent.com/65807151/153704710-e59d95cb-6cfb-427c-bd17-bb11fb3ebc1c.png)

**Reply by PathogenDavid** [reply]
Glad to hear you got things sorted, thanks for coming back to share your solution!

**Comment by the13fools** [comment]
fwiw i think your issue here might have been that you were either missing a ~ or had an extra / in your glob: "/imgui/*.cpp"
